src/ClimaLSM.jl

```diff
@@ -1,5 +1,164 @@
 module ClimaLSM
 
-greet() = print("Hello World!")
+using ClimaCore
+import ClimaCore: Fields
+include("Domains.jl")
+using .Domains
+
+export AbstractModel,
+    make_rhs,
+    make_ode_function,
+    make_update_aux,
+    initialize_prognostic,
+    initialize_auxiliary,
+    initialize_states,
+    LandModel,
+    prognostic_vars,
+    auxiliary_vars
+## Default methods for all models - to be in a seperate module at some point.
+
+abstract type AbstractModel{FT <: AbstractFloat} end
+
+"""
+   prognostic_vars(m::AbstractModel)
+
+Returns the prognostic variable symbols for the model in the form of a tuple.
+"""
+prognostic_vars(m::AbstractModel) = ()
+
+"""
+   auxiliary_vars(m::AbstractModel)
+
+Returns the auxiliary variable symbols for the model in the form of a tuple.
+"""
+auxiliary_vars(m::AbstractModel) = ()
+
+"""
+    make_rhs(model::AbstractModel)
+
+Return a `rhs!` function that updates state variables.
+
+`rhs!` should be compatible with OrdinaryDiffEq.jl solvers.
+"""
+function make_rhs(model::AbstractModel)
+    function rhs!(dY, Y, p, t) end
+    return rhs!
+end
+
+"""
+    make_update_aux(model::AbstractModel)
+
+Return an `update_aux!` function that updates auxiliary parameters `p`.
+"""
+function make_update_aux(model::AbstractModel)
+    function update_aux!(p, Y, t) end
+    return update_aux!
+end
+
+"""
+    make_ode_function(model::AbstractModel)
+
+Returns an `ode_function` that updates auxiliary variables and
+updates the prognostic state.
+
+`ode_function!` should be compatible with OrdinaryDiffEq.jl solvers.
+"""
+function make_ode_function(model::AbstractModel)
+    rhs! = make_rhs(model)
+    update_aux! = make_update_aux(model)
+    function ode_function!(dY, Y, p, t)
+        update_aux!(p, Y, t)
+        rhs!(dY, Y, p, t)
+    end
+    return ode_function!
+end
+
+"""
+    initialize_prognostic(model::AbstractModel,_)
+
+Returns a FieldVector of prognostic variables for `model` with the required
+structure.  This is a stub. We will change it, if possible,
+ as we determine what components require.
+An idea is to use something like `similar(coords)`.
+"""
+function initialize_prognostic(model::AbstractModel, _)
+    return ClimaCore.Fields.FieldVector(; model.model_name => [])
+end
+
+
+"""
+    initialize_auxiliary(model::AbstractModel,_)
+
+Returns a FieldVector of auxiliary variables for `model` with the required
+structure.  This is a stub. We will change it, if possible,
+ as we determine what components require.
+An idea is to use something like `similar(coords)`."""
+function initialize_auxiliary(model::AbstractModel, _)
+    return ClimaCore.Fields.FieldVector(; model.model_name => [])
```

Comment:
is this the best way to not add variables? It would imply that in this case, p.model_name is Any[]. Is it better not to even have `model_name` as part of `p`?

Reply:
My instinct is to not have `model_name` as part of `p`, since a particular `p` already belongs to a particular model, right? The only case were I would want `p.model_name` is a multi-component LSM run.

Reply:
Agreed - we dont really want the hierarchy in a single component run. But I thought we had worked out that we need to pass the entire `p` in the LSM run to each component RHS (due to shared variables). and so for consistency in the RHS function, we also need to pass the entire `p` in single component mode. that is, the arguments to rhs_component! can be `dY,Y,p,t` and we always store variables in Y and p as Y.component, p.component, or we can pass
`dY.component, Y.component,p,t` and just unpack `p`....

Maybe we can give the `p` we pass additional structure? like pass p_component (with no hierarchy) and p_interactions as separate pieces, and in single component mode p_interactions could be `nothing`.

Reply:
https://github.com/CliMA/ClimaLSM.jl/issues/4 - will revisit